src/store/reducers/helper.js

```diff
@@ -74,3 +71,86 @@ export function createComponentMessage(component, props, showAvatar) {
     showAvatar
   });
 }
+
+export const requestSession = (socket) => async function(localID) {
+  return new Promise((resolve, reject) => {
+    socket.emit('session_request', ({ 'session_id': localID }));
+    socket.on('session_confirm', (remoteID) => {
+      console.log(`session_confirm:${socket.id} session_id:${remoteID}`);
+      resolve(remoteID);
+    });
+  })
+}
+
+export function getLocalSession(key) {
+  // Attempt to get local session from sessionStorage
+  const cachedSession = sessionStorage.getItem(key);
+  var session = null;
+  if (cachedSession) {
+    // Found existing session in sessionStorage
+    let parsedSession = JSON.parse(cachedSession)
+    // Format conversation from array of object to immutable Map for use by messages components
+    const formattedConversation = parsedSession.conversation
+      ? Object.values(parsedSession.conversation).map(item => Map(item)).slice()
+      : [];
+    // Check if params is undefined
+    const formattedParams = parsedSession.params
+      ? parsedSession.params
+      : {};
+    //Create a new session to return
+      session = {
+      ...parsedSession,
+      conversation: formattedConversation,
+      params: formattedParams
+    }
+  } else {
+    console.log("No existing local session");
+  }
+  // Returns a formatted session object if any found, otherwise return undefined
+  return session;
+}
+
+export function storeLocalSession(key, sid) {
+  // Attempt to store session id to local sessionStorage
+  const cachedSession = sessionStorage.getItem(key);
+  var session;
+  if (cachedSession) {
+      // Found exisiting session in sessionStorage
+    let parsedSession = JSON.parse(cachedSession)
+    session = {
+      ...parsedSession,
+      session_id: sid
+    }
+  } else {
+    // No existing local session, create a new empty session with only session_id
+    session = {
+      session_id: sid
+    }
+  }
+  // Store updated session to sessionStorage
+  sessionStorage.setItem(key, JSON.stringify(session));
+}
+
+export function storeMessage(conversation) {
+  // Store a conversation List to sessionStorage
+  const localSession = getLocalSession(SESSION_NAME);
+  const newSession = {
+    // Since immutable List is not a native JS object, store conversation as array
+    ...localSession,
+    conversation: [...Array.from(conversation)].slice()
```

Comment:
the spread operator already creates a shallow copy, which is I think the same as slice() method so you don't need it here either

Reply:
Fixed